lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -645,21 +645,28 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
         queryLocalState pt = do
             pp <- localStateQueryQ `send`
                 CmdQueryLocalState pt (QueryIfCurrentShelley Shelley.GetCurrentPParams)
-            handleParamsUpdate fromShelleyPParams pp
+            handleParamsUpdate pt fromShelleyPParams pp
 
             st <- localStateQueryQ `send`
                 CmdQueryLocalState pt (QueryIfCurrentByron Byron.GetUpdateInterfaceState)
-            handleParamsUpdate protocolParametersFromUpdateState st
+            handleParamsUpdate pt protocolParametersFromUpdateState st
 
         handleParamsUpdate
-            :: (p -> W.ProtocolParameters)
+            :: Point (CardanoBlock sc)
+            -> (Maybe Bound -> p -> W.ProtocolParameters)
             -> Either AcquireFailure (Either (MismatchEraInfo (CardanoEras sc)) p)
             -> m ()
-        handleParamsUpdate convert = \case
-            Left (e :: AcquireFailure) ->
-                traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e
-            Right (Right ls) ->
-                    onPParamsUpdate' $ convert ls
+        handleParamsUpdate pt convert = \case
+            Left (e1 :: AcquireFailure) ->
+                traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e1
+            Right (Right ls) -> do
+                    boundRes <- localStateQueryQ `send`
+                        CmdQueryLocalState pt (QueryAnytimeShelley GetEraStart)
+                    case boundRes of
+                        Right boundM ->
+                            onPParamsUpdate' $ convert boundM ls
+                        Left (e2 :: AcquireFailure) ->
+                            traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e2
```

Comment:
Hmmm. The logic feels sort of weird / intertwined here. Query data is mixed with handling results which makes the whole a bit hard to follow (+ causes some logic duplication in the error handling). 

What about:

- run the GetCurrentPParams & GetEraStart in "parallel"
- combine them via an applicative together in `handleParamsUpdate`

So something along the lines of:

```hs
queryLocalState
    :: Point (CardanoBlock sc)
    -> m ()
queryLocalState pt = do
    mb <- localStateQueryQ `send`
        CmdQueryLocalState pt (QueryAnytimeShelley GetEraStart)

    pp <- localStateQueryQ `send`
        CmdQueryLocalState pt (QueryIfCurrentShelley Shelley.GetCurrentPParams)

    sequence (handleParamsUpdate fromShelleyPParams <$> mb <*> pp)
        >>= handleAcquireFailure

    st <- localStateQueryQ `send`
        CmdQueryLocalState pt (QueryIfCurrentByron Byron.GetUpdateInterfaceState)

    sequence (handleParamsUpdate protocolParametersFromUpdateState <$> mb <*> st)
        >>= handleAcquireFailure

handleAcquireFailure
    :: Either AcquireFailure ()
    -> m ()
handleAcquireFailure = \case
    Right () ->
        pure ()
    Left e ->
        traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e

handleParamsUpdate
    :: (Maybe Bound -> p -> W.ProtocolParameters)
    -> (Maybe Bound)
    -> (Either (MismatchEraInfo (CardanoEras sc)) p)
    -> m ()
handleParamsUpdate convert boundM = \case
    Right ls -> do
        onPParamsUpdate' $ convert boundM ls
    Left mismatch -> do
        traceWith tr $ MsgLocalStateQueryEraMismatch mismatch 
```

Reply:
yes, it is much better. I was planning to do it after proof of concept solution

Reply:
done in https://github.com/input-output-hk/cardano-wallet/pull/1932/commits/611ac058eab27fc8eb60898f47bda56bfbb6aba0